lib/scholar/interpolation/cubic_spline.ex

```diff
@@ -0,0 +1,147 @@
+  """
+  defn train(x, y) do
+    # https://en.wikiversity.org/wiki/Cubic_Spline_Interpolation
+    # Reference implementation in Scipy
+
+    x = Nx.as_type(x, :f64)
+    y = Nx.as_type(y, :f64)
+
+    dx = x[1..-1//1] - x[0..-2//1]
+    n = Nx.size(x)
+
+    slope = (y[1..-1//1] - y[0..-2//1]) / dx
+
+    s =
+      if n == 3 do
+        a =
+          Nx.concatenate(
+            [
+              Nx.tensor([[1, 1, 0]]),
+              Nx.stack([dx[1], 2 * (dx[0] + dx[1]), dx[0]]) |> Nx.new_axis(0),
+              Nx.tensor([[0, 1, 1]])
+            ],
+            axis: 0
+          )
+
+        b = Nx.stack([2 * slope[0], 3 * (dx[0] * slope[1] + dx[1] * slope[0]), 2 * slope[1]])
+
+        Nx.LinAlg.solve(a, b)
+      else
+        # n > 3 and bc = not_a_knot
+
+        d_0 = Nx.new_axis(x[2] - x[0], 0)
+        d_n = Nx.new_axis(x[-1] - x[-3], 0)
+
+        up_diag =
+          Nx.concatenate([
+            d_0,
+            dx[0..-2//1]
+          ])
+
+        low_diag =
+          Nx.concatenate([
+            dx[1..-1//1],
+            d_n
+          ])
+
+        main_diag =
+          Nx.concatenate([
+            Nx.new_axis(dx[1], 0),
+            2 * (dx[0..-2//1] + dx[1..-1//1]),
+            Nx.new_axis(dx[-2], 0)
+          ])
+
+        a =
+          Nx.broadcast(0, {n, n})
+          |> Nx.put_diagonal(up_diag, offset: 1)
+          |> Nx.put_diagonal(main_diag, offset: 0)
+          |> Nx.put_diagonal(low_diag, offset: -1)
+
+        bc_0 = ((dx[0] + 2 * d_0) * dx[1] * slope[0] + dx[0] ** 2 * slope[1]) / d_0
+
+        bc_n = (dx[-1] ** 2 * slope[-2] + (2 * d_n + dx[-1]) * dx[-2] * slope[-1]) / d_n
+
+        b =
+          Nx.concatenate([
+            Nx.new_axis(bc_0, 0),
+            3 * (dx[1..-1//1] * slope[0..-2//1] + dx[0..-2//1] * slope[1..-1//1]),
+            Nx.new_axis(bc_n, 0)
+          ])
+
+        Nx.LinAlg.solve(a, b)
+      end
+
+    slope = (y[1..-1//1] - y[0..-2//1]) / dx
+    t = (s[0..-2//1] + s[1..-1//1] - 2 * slope) / dx
+
+    c_3 = t / dx
+    c_2 = (slope - s[0..-2//1]) / dx - t
+    c_1 = s[0..-2//1]
+    c_0 = y[0..-2//1]
+
+    c = Nx.stack([c_3, c_2, c_1, c_0], axis: 0)
+
+    %__MODULE__{coefficients: c, x: x}
+  end
+
+  @doc """
+  Returns the value fit by `train/2` corresponding to the `target_x` input
+
+  ### Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+  """
+  deftransform predict(
+                 %__MODULE__{} = model,
+                 target_x,
+                 opts \\ []
+               ) do
+    predict_n(model, target_x, NimbleOptions.validate!(opts, @opts_schema))
+  end
+
+  defnp predict_n(%__MODULE__{x: x, coefficients: coefficients}, target_x, opts \\ []) do
+    nan_selector =
+      Nx.logical_and(opts[:extrapolate] == false, target_x < x[0] or target_x > x[-1])
+
+    nan = Nx.tensor(:nan, type: :f64)
+    target_x = Nx.as_type(target_x, :f64)
+
+    idx =
+      Nx.sum(x < Nx.new_axis(target_x, 1), axes: [1])
```

Comment:
It may be just me, but I think broadcasting tensor rank is confusing and adds overhead when reading the code, so I favor reshaping both operands.

```suggestion
      (Nx.new_axis(target_x, 1) > Nx.new_axis(x, 0))
      |> Nx.sum(axes: [1])
```

Sidenote: instead of `Nx.sum` we could use `Nx.argmax`, but I don't see a clear reason why either should be more efficient :)

Reply:
argmax ended up being the fix I needed for an off-by-one